x-pack/test/functional/page_objects/security_page.ts

```diff
@@ -17,53 +18,114 @@ export function SecurityPageProvider({ getService, getPageObjects }) {
   const userMenu = getService('userMenu');
   const PageObjects = getPageObjects(['common', 'header', 'settings', 'home', 'error']);
 
-  class LoginPage {
-    async login(username, password, options = {}) {
-      const [superUsername, superPassword] = config.get('servers.elasticsearch.auth').split(':');
+  interface LoginOptions {
+    expectSpaceSelector?: boolean;
+    expectSuccess?: boolean;
+    expectForbidden?: boolean;
+  }
+
+  type LoginExpectedResult = 'spaceSelector' | 'error' | 'chrome';
+
+  async function waitForLoginPage() {
+    await retry.waitForWithTimeout('login page', config.get('timeouts.waitFor') * 5, async () => {
+      const alert = await browser.getAlert();
+      if (alert && alert.accept) {
+        await alert.accept();
+      }
+      return await find.existsByDisplayedByCssSelector('.login-form');
+    });
+  }
+
+  async function waitForLoginForm() {
```

Comment:
Do we need to distinguish between waiting for the login form vs the login page? Can we collapse these into a single function?

Reply:
> Do we need to distinguish between waiting for the login form vs the login page?

I believe so, yes. I mean Login Form isn't rendered by default when we have multiple providers and Login Selector isn't rendered when we have just basic/token. And after logout we don't know where we end up (it's used in various tests, with or without Login Selector) so we're waiting for Login Page to appear.

The `login-form` CSS class we use for the entire Login Page is rather confusing though.

Or what did you mean?

Reply:
If I understand the distinction correctly, we have:
1) Login form, which is the username/password input fields
2) Login page, which is the screen that contains the Login form (could alternatively render the disabled login view if we are preventing users from logging in)
3) Login selector, which is the new one

To me, it appears like we don't need to make a distinction between Login form and Login page, since the form is just a child of the page, and not a distinct view.

Is there something I'm overlooking or oversimplifying?

Reply:
>1.  Login form, which is the username/password input fields

Correct

> 2. Login page, which is the screen that contains the Login form (could alternatively render the disabled login view if we are preventing users from logging in)

It _may_ contain Login Form (including its disabled states) _or_ it _may_ contain Login Selector (we render just one or another)

> 3. Login selector, which is the new one

Correct, it's alternative to the Login Form that Login Page _may_ contain

> To me, it appears like we don't need to make a distinction between Login form and Login page, since the form is just a child of the page, and not a distinct view.
> Is there something I'm overlooking or oversimplifying?

Let me explain two possible cases when I _think_ we need two different things so that you can point to the missing piece in my understanding:

* if some test call `forceLogout`, we should wait for either Login Form or Login Selector to be rendered. But the problem is that we don't know (unless we start analyzing Kibana server config) if Kibana is configured with basic only (in this case `Login Page -> Login Form` will be rendered) or it's configured with more than one provider (in this case `Login Page -> Login Selector` will be rendered). Waiting for parent Login Page seems to do the job in both cases.

* when we test Login Page itself we want to know that transition between Login Selector and Login Form works, so we have two different methods that wait for one or another to become visible (parent Login Page is always visible/rendered in these tests).

How would you simply this to cover both cases with just two methods? I'm all in to simply stuff as much as possible, I'm just not sure I'm completely following your proposal, sorry about that :see_no_evil:

Reply:
Ah, the piece I was missing was the the Login Page contains the Login Selector. That makes sense, so I think we need to keep both here. Thanks for clarifying this for me!